<a href="https://colab.research.google.com/github/rominaeh/datascience-with-R/blob/main/statistical_inference.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [6]:
# مثال 6
# confidence interval
# manually calculating
(3/sqrt(25))*qnorm(.975)
d = round((3/sqrt(25))*qnorm(.975),1)
d
c(50-d , 50+d)
print('--------------------------')
ci.mu.test = function(xbar, sigma, n, alpha=0.05){
c(xbar-sigma/sqrt(n)*qnorm(1-alpha/2), xbar+sigma/sqrt(n)*qnorm(1-alpha/2))}
ci.mu.test(50, 3, 25)

[1] 1.175978

[1] 1.2

[1] 48.8 51.2

[1] "--------------------------"


[1] 48.82402 51.17598

In [9]:
# مثال 7
# with t test calculating
#  for the mean of a sample
battery = c(21, 24, 18, 17, 20, 26, 28, 23, 21)
t.test(battery, mu=22)$conf.int
# for the variance of the sample - manually
ci.var.test = function(x, alpha = 0.05){
    n = length(x)
    c((n-1)*var(x)/qchisq(1-alpha/2, n-1), (n-1)*var(x)/qchisq(alpha/2, n-1))
}
ci.var.test(battery)

[1] 19.22853 24.77147
attr(,"conf.level")
[1] 0.95

[1]  5.931149 47.712315

In [10]:
# مثال 8
# with binom test calculating
binom.test(44,100,p=0.5)$conf.int
# (x,sample,alpha)
# manually
c(0.44-sqrt(0.44*0.56/100)*qnorm(0.975), 0.44+ sqrt(0.44*0.56/100)*qnorm(0.975))

[1] 0.3408360 0.5428125
attr(,"conf.level")
[1] 0.95

[1] 0.3427099 0.5372901

In [12]:
# مثال 12
# one sample t test

ERROR: ignored

In [ ]:
# مثال 17
# corr test

In [ ]:
# مثال 18
# f test

In [ ]:
# مثال 21
# two sample t test

In [ ]:
# مثال 22
# one way ANOVA


In [ ]:
# مثال 25
# Fisher's LSD Test
    # After conducting ANOVA, if significant differences are found among group means, the LSD test helps determine which specific groups differ.
    # The LSD test compares the means of each pair of groups and determines if the difference is statistically significant.
    # The test involves comparing the absolute difference between the means of two groups against a critical value.
wheat= c(7,7, 15,11, 9,12, 17,12, 18,18, 14, 18, 18, 19, 19,19, 25,22, 19,23, 7,10, 11,15,11)
nitrate = factor(rep(c(1:5), each=5))
wheatdf = data.frame(wheat, nitrate)
wheatdf

wheat nitrate
1   7    1      
2   7    1      
3  15    1      
4  11    1      
5   9    1      
6  12    2      
7  17    2      
8  12    2      
9  18    2      
10 18    2      
11 14    3      
12 18    3      
13 18    3      
14 19    3      
15 19    3      
16 19    4      
17 25    4      
18 22    4      
19 19    4      
20 23    4      
21  7    5      
22 10    5      
23 11    5      
24 15    5      
25 11    5

In [ ]:
wheataov = aov(wheat ~ nitrate, wheatdf)
wheataov

Call:
   aov(formula = wheat ~ nitrate, data = wheatdf)

Terms:
                nitrate Residuals
Sum of Squares   475.76    161.20
Deg. of Freedom       4        20

Residual standard error: 2.839014
Estimated effects may be unbalanced

    nitrate: This represents the variance in the wheat measurements that can be attributed to the different levels of nitrate.
    Residuals: This represents the variance in the wheat measurements that is not explained by the nitrate levels.

Sum of Squares:

    475.76 for nitrate: This is the total variance in wheat explained by the different levels of nitrate.
    161.20 for Residuals: This is the variance in wheat that cannot be explained by the nitrate levels (i.e., it's the unexplained or error variance).

Degrees of Freedom:

    4 for nitrate: This indicates that there are 5 (4+1) levels of nitrate treatments.
    20 for Residuals: This is calculated as the total number of observations minus the number of levels in the nitrate variable.

Residual Standard Error:

    2.839014: This is the standard deviation of the residuals. It measures the typical distance between the observed wheat values and the values predicted by the model. A lower residual standard error indicates a better fit of the model to the data.

Estimated Effects May Be Unbalanced:

    This note suggests that the data may not be balanced, meaning that the groups (levels of nitrate) might not have equal sample sizes. Unbalanced data can affect the interpretation and the statistical power of ANOVA.

If the ANOVA results indicate significant differences among the group means (e.g., significant sum of squares for the treatment or factor, as in your case with the nitrate variable), it suggests that not all group means are equal. However, ANOVA doesn't tell you which specific groups are different from each other.

    You conducted an ANOVA (Analysis of Variance) to assess whether there are significant differences in the wheat measurements across different levels of nitrate.
    The ANOVA results indicated a significant variance in wheat yields attributable to the different levels of nitrate (as evidenced by a high sum of squares for the nitrate factor). This suggests that not all levels of nitrate have the same effect on wheat yields.

In [ ]:
install.packages('agricolae'); update.packages('agricolae')
library('agricolae')
with(wheataov, pairwise.t.test(wheat, nitrate, p.adjust="none")) #hough pairwise.t.test and lst.test() are used for making pairwise comparisons between group means after an ANOVA.

Installing package into ‘/usr/local/lib/R/site-library’
(as ‘lib’ is unspecified)




	Pairwise comparisons using t tests with pooled SD 

data:  wheat and nitrate 

  1       2       3       4      
2 0.00541 -       -       -      
3 0.00031 0.23471 -       -      
4 2.1e-06 0.00251 0.03754 -      
5 0.58375 0.01859 0.00116 7.0e-06

P value adjustment method: none 

    Nitrate Levels 1 and 2 (p = 0.00541): There is a significant difference between these levels.
    Nitrate Levels 1 and 3 (p = 0.00031): There is a significant difference.
    Nitrate Levels 1 and 4 (p = 2.1e-06): There is a significant difference.
    Nitrate Levels 1 and 5 (p = 0.58375): No significant difference.
    Nitrate Levels 2 and 3 (p = 0.23471): No significant difference.
    Nitrate Levels 2 and 4 (p = 0.00251): There is a significant difference.
    Nitrate Levels 2 and 5 (p = 0.01859): There is a significant difference.
    Nitrate Levels 3 and 4 (p = 0.03754): There is a significant difference.
    Nitrate Levels 3 and 5 (p = 0.00116): There is a significant difference.
    Nitrate Levels 4 and 5 (p = 7.0e-06): There is a significant difference.

P Value Adjustment Method: None:

    This indicates that the p-values are not adjusted for multiple comparisons. In the context of LSD tests, this is standard practice, but it's important to be aware that it can increase the risk of Type I errors (false positives), especially with many comparisons.

In [ ]:
# مثال 28
# ANOVA test
f1 = factor(c(1, 1, 1, 1, 2, 2, 2, 2)) # Factor for type of car
f2 = factor(c(1, 1, 2, 2, 1, 1, 2, 2)) # Factor for type of gasoline
y = c(26.7, 25.2, 28.6, 29.3, 32.3, 32.8, 26.1, 24.2) # Miles per gallon data
ydf = data.frame(y, f1, f2); # Creating a data frame
ydf
#  2 types of cars with 2 type of gasoline each how much gas they consume

y    f1 f2
1 26.7 1  1 
2 25.2 1  1 
3 28.6 1  2 
4 29.3 1  2 
5 32.3 2  1 
6 32.8 2  1 
7 26.1 2  2 
8 24.2 2  2

In [ ]:
summary(aov(y ~ f1 * f2))

            Df Sum Sq Mean Sq F value  Pr(>F)   
f1           1   3.92    3.92   4.752 0.09477 . 
f2           1   9.68    9.68  11.733 0.02665 * 
f1:f2        1  54.08   54.08  65.552 0.00126 **
Residuals    4   3.30    0.82                   
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

    Factor A (Gasoline Type): With a p-value of 0.09477, it's statistically significant at the 10% level but not at the 5% level, suggesting a marginal effect of gasoline type on miles per gallon.
    Factor B (Car Type): This factor is significant (p < 0.05), indicating a clear effect of car type on miles per gallon.
    Interaction A:B (Interaction between Car Type and Gasoline Type): The interaction is highly significant (p < 0.05), suggesting that the effect of one factor depends on the level of the other factor.

    Sum of Squares (Sum Sq):This is a measure of the total variation in the data. It's calculated as the sum of squared differences from the mean.
    For each factor (like f1 and f2), the Sum of Squares tells us how much of the total variation in the data is due to that factor. The Residual Sum of Squares represents the variation that is not explained by the model.

    Mean Square (Mean Sq):

    The Mean Square is calculated by dividing the Sum of Squares by the corresponding degrees of freedom.
    It's an average of the squared deviations and provides an estimate of variance for each factor and for the residuals.

    F value:

    The F value is a test statistic used to determine whether there is a significant difference between groups.
    It is calculated by dividing the Mean Square of the factor by the Mean Square of the residuals. A larger F value indicates a more significant difference between groups

In [ ]:
LSD.test(wheataov, 'nitrate', console=T)


Study: wheataov ~ "nitrate"

LSD t Test for wheat 

Mean Square Error:  8.06 

nitrate,  means and individual ( 95 %) CI

  wheat      std r       se       LCL      UCL Min Max Q25 Q50 Q75
1   9.8 3.346640 5 1.269646  7.151566 12.44843   7  15   7   9  11
2  15.4 3.130495 5 1.269646 12.751566 18.04843  12  18  12  17  18
3  17.6 2.073644 5 1.269646 14.951566 20.24843  14  19  18  18  19
4  21.6 2.607681 5 1.269646 18.951566 24.24843  19  25  19  22  23
5  10.8 2.863564 5 1.269646  8.151566 13.44843   7  15  10  11  11

Alpha: 0.05 ; DF Error: 20
Critical Value of t: 2.085963 

least Significant Difference: 3.745452 

Treatments with the same letter are not significantly different.

  wheat groups
4  21.6      a
3  17.6      b
2  15.4      b
5  10.8      c
1   9.8      c


    Mean Square Error:
        Mean Square Error: 8.06: This is the mean of the squared differences between each observation and the overall mean (residuals). It's a measure of the variance within the groups.

    Group Mean Information:
        The table lists each nitrate group (1 to 5), their mean wheat yields (wheat), standard deviations (std), number of replicates (r), lower and upper confidence limits (LCL and UCL), and the range (Min and Max).
        For example, for group 1, the mean wheat yield is 9.8 with a standard deviation of 3.346640, and the 95% confidence interval for the mean is between 7.151566 and 12.44843.

    Statistical Parameters:
        alpha: 0.05: The significance level used for the test.
        Df Error: 20: Degrees of freedom for the error term.
        Critical Value of t: 2.085963: The t-value from the t-distribution used for significance testing at the given alpha level.

    Least Significant Difference:
        Least Significant Difference 3.745452: This value is the minimum difference between means required for them to be considered significantly different. If the difference between any two means is greater than this value, it is considered statistically significant.

    Group Comparisons and Classifications:
        The groups are then classified based on whether their means differ significantly. Groups with means not significantly different from each other are assigned the same letter.
        In your result, group 4 (with mean 21.6) is significantly different from all others and is labeled 'a'. Groups 3 and 2 are labeled 'b', indicating that they are not significantly different from each other but are different from group 4. Groups 5 and 1 are labeled 'c', indicating they are not significantly different from each other but are different from groups 2, 3, and 4.

This output is useful for understanding not only the mean values of wheat yield at different nitrate levels but also how these means compare statistically. It helps in identifying which specific nitrate levels lead to significantly different wheat yields.

It directly compares the means of each pair of groups and uses a t-distribution to determine if differences are significant

In [ ]:
# تمرین 1
# conducting statistical tests on a dataset representing your monthly telephone expenses over a 12-month period. The data points provided are as follows: 46, 33, 39, 37, 46, 30, 48, 32, 49, 35, 30, 48.

# There are two parts to the question:

#     Test whether the mean of the total amount paid is equal to 44:
#     This involves performing a one-sample t-test. The null hypothesis (H0) is that the mean is 44, and the alternative hypothesis (H1) is that the mean is not 44.

#     Test whether the variance of the total amount paid is equal to 60:
#     For this, a chi-squared test for variance is used. The null hypothesis (H0) is that the variance is 60, while the alternative hypothesis (H1) is that the variance is not 60.




	One Sample t-test

data:  data
t = -2.1065, df = 11, p-value = 0.05894
alternative hypothesis: true mean is not equal to 44
95 percent confidence interval:
 34.62771 44.20563
sample estimates:
mean of x 
 39.41667 



In [ ]:
data <- c(46, 33, 39, 37, 46, 30, 48, 32, 49, 35, 30, 48)
t_test_result <- t.test(data, mu=44)
print(t_test_result)


	One Sample t-test

data:  data
t = -2.1065, df = 11, p-value = 0.05894
alternative hypothesis: true mean is not equal to 44
95 percent confidence interval:
 34.62771 44.20563
sample estimates:
mean of x 
 39.41667 



    t-value: The t-value, -2.1065 in your case, is a measure of how many standard deviations the sample mean (39.41667) is from the hypothesized mean (44). A negative t-value indicates that the sample mean is less than the hypothesized mean.

    Degrees of Freedom (df): The degrees of freedom, 11 in this case, are calculated as the number of observations minus 1 (12 - 1 = 11). This is used to determine the critical value from the t-distribution that the t-value should be compared against.

    p-value: The p-value of 0.05894 tells you the probability of observing a result as extreme as, or more extreme than, the one observed (t = -2.1065) if the null hypothesis (that the mean is 44) were true. In conventional hypothesis testing, a p-value less than 0.05 is often considered statistically significant. Here, the p-value is slightly above 0.05, which suggests that the evidence against the null hypothesis is not strong enough to conclude that the mean is significantly different from 44.

    95% Confidence Interval: This interval (34.62771 to 44.20563) represents the range of values within which you can be 95% confident that the true mean of the population lies. Since 44 (the hypothesized mean) falls within this interval, it suggests that there is not strong evidence against the hypothesis that the mean is 44.

    Sample Estimates: The mean of x (39.41667) is the sample mean of your data. It's the average of the values you provided and is the point estimate of the population mean.

In statistical hypothesis testing, we don't exactly "accept" the null hypothesis (H0); rather, we either reject it or fail to reject it based on the evidence provided by our data.

In your case, the null hypothesis (H0) is that the true mean of the phone expenses is 44. Given that your p-value (0.05894) is greater than the conventional threshold of 0.05, you do not have sufficient evidence to reject this hypothesis. This means you fail to reject the null hypothesis.

However, "failing to reject" the null hypothesis is not the same as proving it true or accepting it. It simply means that your data does not provide strong enough evidence against H0. In other words, based on your data, it's plausible that the true mean is 44, but you're not confirming this with certainty.

So, in summary, your result indicates that while there is some indication that the mean might not be 44 (as seen from the sample mean of 39.41667), this evidence is not strong enough to confidently state that the true mean differs from 44. You do not reject H0, but this conclusion is not a strong endorsement of H0 either. It's more about the lack of sufficient evidence to prove otherwise.

In [ ]:
n <- length(data)
sample_variance <- var(data)
chi_sq_statistic <- (n-1) * sample_variance / 60
p_value_variance <- pchisq(chi_sq_statistic, df=n-1, lower.tail=FALSE)
print(chi_sq_statistic) #measures how much your observed variance deviates from the hypothesized variance
print(p_value_variance) #how extreme the chi-squared statistic is, assuming the null hypothesis is true

[1] 10.41528
[1] 0.4934642


there isn't enough evidence to reject the null hypothesis that the true variance of your phone expenses is 60.